GitHub.Collectors/Processor/PointProcessor.cs

```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.CloudMine.GitHub.Collectors.Processor
+{
+    public class PointProcessor
+    {
+        private readonly CollectorBase<GitHubCollectionNode> collector;
+
+        private readonly List<IRecordWriter> recordWriters;
+        private readonly ICache<PointCollectorTableEntity> cache;
+        private readonly ITelemetryClient telemetryClient;
+        private readonly string apiDomain;
+
+        public PointProcessor(IAuthentication authentication,
+                              List<IRecordWriter> recordWriters,
+                              GitHubHttpClient httpClient,
+                              ICache<PointCollectorTableEntity> cache,
+                              ITelemetryClient telemetryClient,
+                              string apiDomain)
+        {
+            this.collector = new GitHubCollector(httpClient, authentication, telemetryClient, recordWriters);
+            this.recordWriters = recordWriters;
+            this.cache = cache;
+            this.telemetryClient = telemetryClient;
+            this.apiDomain = apiDomain;
+        }
+
+        public async Task ProcessAsync(PointCollectorInput input)
+        {
+            Dictionary<string, JToken> additionalMetadata = new Dictionary<string, JToken>();
+            JObject context = input.Context;
+            if (context != null)
+            {
+                foreach (KeyValuePair<string, JToken> contextElement in context)
+                {
+                    additionalMetadata.Add(contextElement.Key, contextElement.Value);
+                }
+            }
+    
+            GitHubCollectionNode collectionNode = new GitHubCollectionNode()
+            {
+                RecordType = input.RecordType,
+                ApiName = input.ApiName,
+                GetInitialUrl = metadata => input.Url,
+                AdditionalMetadata = additionalMetadata,
+            };
+
+            foreach (IRecordWriter recordWriter in this.recordWriters)
+            {
+                await recordWriter.NewOutputAsync(input.RecordType).ConfigureAwait(false);
+            }
+            await this.ProcessAndCacheBatchingRequestAsync(input, collectionNode).ConfigureAwait(false);
+        }
+
+        private async Task ProcessAndCacheBatchingRequestAsync(PointCollectorInput input, GitHubCollectionNode collectionNode)
+        {
+            string apiName = collectionNode.ApiName;
+            if (!input.IgnoreCache && !input.IgnoreCacheForApis.Contains(apiName))
+            {
+                // PointCollectorTableEntity tableEntity = await this.cache.RetrieveAsync(new PointCollectorTableEntity(input));
+                // TODO Logic for skipping if in cache goes here
+                bool apiCollected = false;
+                if (apiCollected)
+                {
+                    // TODO track event in telemetry
+                    return;
```

Comment:
Are you planning to implement these before merging, or do you plan to bring those in a separate PR? Regardless, as I mentioned above, I don't think this cache is working as you hope for. 

If I recall our discussion correctly, we want to have a cache because we would like to ensure that we are not doing point collection on the same URL over and over, especially over a short amount of time. Here is the concrete problem:
* microsoft organization does updates to 1000 PRs.
* GH fires 1000 events where organization URL is: github.com/microsoft
* We add 1000 messages to the point collector queue with the same request.

I think the idea is that you don't even want to add these messages at the beginning. You can do this only by keeping a cache that tells you when you have collected a particular URL latest (this to be updated / inserted by the point collector) and then not redirecting the collection to the point collector if this URL was collected recently (let's say within the last 5 minutes). This check and skip needs to be baked into Webhook processor.

In other words, point collector would be a producer for this cache whereas the webhook processor would be a consumer.

If you had something else in mind, please let me know. I might be mis-remembering.